src/Simulation/Simulators/OpenSystemsSimulator/DataModel/Generator.cs

```diff
@@ -0,0 +1,146 @@
+) : Generator(NQubits);
+
+public class GeneratorConverter : JsonConverter<Generator>
+{
+    public override Generator Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+    {
+        reader.Require(JsonTokenType.StartObject, readCurrent: false);
+        var arrayConverter = new ComplexArrayConverter();
+        return reader.ReadQubitSizedData<Generator>((ref Utf8JsonReader reader, string kind) =>
+        {
+            switch (kind)
+            {
+                case "Unsupported":
+                    return (int nQubits) => new UnsupportedGenerator(nQubits);
+
+                case "ExplicitEigenvalueDecomposition":
+                    NDArray? eigenvectors = null;
+                    NDArray? eigenvalues = null;
+
+                    while (reader.Read())
+                    {
+                        if (reader.TokenType == JsonTokenType.EndObject)
+                        {
+                            // We're at the end of the object, and can break out of the
+                            // read loop.
+                            break;
+                        }
+
+                        // If it's not the end of the object, the current token needs
+                        // to be a property name.
+                        if (reader.TokenType != JsonTokenType.PropertyName)
+                        {
+                            throw new JsonException();
+                        }
+
+                        var propertyName = reader.GetString();
+                        reader.Read();
+
+                        switch (propertyName)
+                        {
+                            case "vectors":
+                                eigenvectors = arrayConverter.Read(
+                                    ref reader, typeof(NDArray), options
+                                );
+                                break;
+
+                            case "values":
+                                eigenvalues = arrayConverter.Read(
+                                    ref reader, typeof(NDArray), options
+                                );
+                                break;
+
+                            default:
+                                throw reader.CurrentState.Exception($"Unknown property {propertyName}.");
+                        }
+
+                    }
+
+                    if (eigenvalues == null)
+                    {
+                        throw reader.CurrentState.Exception($"Generator kind was ExplicitEigenvalueDecomposition, but no eigenvalues were provided.");
+                    }
+                    Debug.Assert(eigenvalues is not null);
+
+                    if (eigenvectors == null)
+                    {
+                        throw reader.CurrentState.Exception($"Generator kind was ExplicitEigenvalueDecomposition, but no eigenvectors were provided.");
+                    }
+                    Debug.Assert(eigenvectors is not null);
+
```

Comment:
I have an impression that in both cases the `Debug.Assert()` is redundant after the corresponding `if`.
```suggestion
                    if (eigenvalues == null)
                    {
                        throw reader.CurrentState.Exception($"Generator kind was ExplicitEigenvalueDecomposition, but no eigenvalues were provided.");
                    }

                    if (eigenvectors == null)
                    {
                        throw reader.CurrentState.Exception($"Generator kind was ExplicitEigenvalueDecomposition, but no eigenvectors were provided.");
                    }
```

Reply:
Here as well, the extraneous asserts are used to suppress warnings in C# by providing the compiler with appropriate metadata to prove that neither `eigenvectors` nor `eigenvalues` is `null` in the case that no exception is thrown.

Reply:
Hm.... OK. I thought the `if` with nested `throw` was enough to convince the compiler that after the `if` the var is not `null`.

Reply:
It almost always is, agreed; I'm honestly not sure why Roslyn failed to conclude nonnull in this case.